toolkit/featureforms/src/main/java/com/arcgismaps/toolkit/featureforms/components/text/FormTextFieldState.kt

```diff
@@ -85,84 +107,257 @@ internal class FormTextFieldState(
+                    wasFocused = true
                 }
+                updateValidation(value.value)
             }
         }
     }
-
-    /**
-     * Validates the current [value]'s length based on the [minLength], [maxLength], and [isRequired] and sets the
-     * [hasError] and [_errorMessage] if there was an error in validation.
-     */
-    private fun validate(value: String) {
-        _hasError.value = if (isRequired.value && value.isEmpty()) {
-            _errorMessage = context.getString(R.string.required)
-            true
-        } else if (value.length !in minLength..maxLength) {
-            _errorMessage = helperText
-            true
+    
+    private fun updateValidation(value: String) {
+        validationErrors = validate(value)
+        val errorToDisplay = errorMessageToDisplay(value, validationErrors)
+        _supportingTextIsErrorMessage.value = errorToDisplay != NoError
+        _supportingText.value = if (errorToDisplay != NoError) {
+            errors[errorToDisplay] ?: throw IllegalStateException("validation error must have a message")
         } else {
-            false
+            description.ifEmpty {
+                if (_isFocused.value) helperText else ""
+            }
         }
+        _hasError.value = validationErrors.isNotEmpty()
     }
-
+    
+    private fun validateTextRange(value: String): ValidationErrorState =
+        if (value.length !in minLength..maxLength) {
+            if (minLength > 0 && maxLength > 0) {
+                if (minLength == maxLength) {
+                    ExactCharConstraint
+                } else {
+                    MinMaxCharConstraint
+                }
+            } else {
+                MaxCharConstraint
+            }
+        } else {
+            NoError
+        }
+    
+    private fun validateNumericRange(value: String): ValidationErrorState =
+        if (domain != null && domain is RangeDomain) {
+            if (fieldType.isIntegerType) {
+                val (min, max) = domain.asLongTuple
+                val numberVal = value.toLong()
+                if (min != null && max != null) {
+                    if (numberVal in min..max) {
+                        NoError
+                    } else {
+                        MinMaxNumericConstraint
+                    }
+                } else if (min != null) {
+                    if (min <= numberVal) {
+                        NoError
+                    } else {
+                        MinNumericConstraint
+                    }
+                    
+                } else if (max != null) {
+                    if (numberVal <= max) {
+                        NoError
+                    } else {
+                        MaxNumericConstraint
+                    }
+                } else {
+                   NoError
+                }
+            } else {
+                val (min, max) = domain.asDoubleTuple
+                val numberVal = value.toDouble()
```

Comment:
```suggestion
                val numberVal = value.toDoubleOrNull()
```
Similarly, here too.